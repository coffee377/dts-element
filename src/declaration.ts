import {Element} from './element';
import {line_map} from './helpers/line-map';

export interface IDeclarationRequiredParameters {
  name: string;
}

export interface IDeclarationOptionalParameters {
  jsdoc: string;
}

export type AnyDeclaration = Declaration<any, any>;

export abstract class Declaration<RequiredParameters extends {}, OptionalParameters extends {}> extends Element
    <IDeclarationRequiredParameters & RequiredParameters, IDeclarationOptionalParameters & OptionalParameters> {

  public get default_declaration_parameters(): IDeclarationOptionalParameters {
    return {
      jsdoc: '',
    };
  }

  public emit_jsdoc(): string {
    const {jsdoc} = this.parameters;
    const content = line_map(jsdoc, (line: string) => ` * ${line}`);
    return (content.length === 0)
      ? content
      : `/**\n${content}\n */\n`;
  }

}
