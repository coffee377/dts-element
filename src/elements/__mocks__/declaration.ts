import {Element} from '../../element';

export abstract class Declaration extends Element<any, any> {

  public get default_declaration_parameters(): any {
    return {};
  }

  public get jsdoc(): string {
    return '[JSDoc]\n';
  }

}
