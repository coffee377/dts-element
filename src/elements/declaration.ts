import {AnyElement, Element} from '../element';
import {emit_jsdoc} from '../helpers/emit-jsdoc';

export interface IDeclarationRequiredParameters {
  name: string;
}

export interface IDeclarationOptionalParameters {
  jsdoc: string;
}

export type AnyDeclaration = Declaration<any, any>;

export abstract class Declaration<RequiredParameters extends {}, OptionalParameters extends {}> extends Element
    <IDeclarationRequiredParameters & RequiredParameters, IDeclarationOptionalParameters & OptionalParameters> {

  // istanbul ignore next
  public get default_declaration_parameters(): IDeclarationOptionalParameters {
    return {
      jsdoc: '',
    };
  }

  public _emit(container: AnyElement | null): string {
    return `${this._emit_jsdoc()}${this._emit_raw(container)}`;
  }

  public _emit_jsdoc(): string {
    return emit_jsdoc(this.parameters.jsdoc);
  }

  public abstract _emit_raw(container: AnyElement | null): string;

}
