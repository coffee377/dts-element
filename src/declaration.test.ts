import { Declaration } from './declaration';

describe('#emit_jsdoc()', () => {
  it('should return formatted jsdoc', () => {
    const declaration = { parameters: { jsdoc: 'line 1\nline 2\nline 3' } };
    expect(Declaration.prototype.emit_jsdoc.call(declaration)).toMatchSnapshot();
  });
});
