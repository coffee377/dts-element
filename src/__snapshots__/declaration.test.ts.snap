// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit_jsdoc() should return formatted jsdoc 1`] = `
"/**
 * line 1
 * line 2
 * line 3
 */
"
`;
